src/vm/checker/check_db.rs

```diff
@@ -0,0 +1,113 @@
+use rusqlite::{Connection, OptionalExtension, NO_PARAMS, Row};
+use rusqlite::types::ToSql;
+
+
+use vm::types::TypeSignature;
+use vm::checker::errors::{CheckError, CheckErrors, CheckResult};
+use vm::checker::typecheck::{ContractAnalysis, FunctionType};
+
+const SQL_FAIL_MESSAGE: &str = "PANIC: SQL Failure in contract analysis.";
+
+pub struct AnalysisDatabase {
+    conn: Connection
+}
+
+impl AnalysisDatabase {
+    pub fn initialize(filename: &str) -> AnalysisDatabase {
+        let mut contract_db = AnalysisDatabase::inner_open(filename);
+        // this is the _laziest_ of structures at the moment.
+        //    more to come!
+        contract_db.execute("CREATE TABLE IF NOT EXISTS type_analysis_table
+                      (contract_identifier INTEGER PRIMARY KEY AUTOINCREMENT,
+                       contract_name TEXT NOT NULL UNIQUE,
+                       analysis TEXT NOT NULL)",
+                            NO_PARAMS);
+
+        contract_db.check_schema();
+
+        contract_db
+    }
+
+    pub fn memory() -> AnalysisDatabase {
+        AnalysisDatabase::initialize(":memory:")
+    }
+
+    pub fn open(filename: &str) -> AnalysisDatabase {
+        let contract_db = AnalysisDatabase::inner_open(filename);
+
+        contract_db.check_schema();
+        contract_db
+    }
+
+    pub fn check_schema(&self) {
+        let sql = "SELECT sql FROM sqlite_master WHERE name=?";
+        let _: String = self.conn.query_row(sql, &["type_analysis_table"],
+                                            |row| row.get(0))
+            .expect("Bad schema in analysis db initialization.");
+    }
+
+    pub fn inner_open(filename: &str) -> AnalysisDatabase {
+        let conn = Connection::open(filename)
+            .expect("Failure to open analysis db.");
+        AnalysisDatabase {
+            conn: conn
+        }
+    }
+
+    pub fn execute<P>(&mut self, sql: &str, params: P) -> usize
+    where
+        P: IntoIterator,
+        P::Item: ToSql {
+        self.conn.execute(sql, params)
+            .expect(SQL_FAIL_MESSAGE)
+    }
+
+    fn query_row<T, P, F>(&self, sql: &str, params: P, f: F) -> Option<T>
+    where
+        P: IntoIterator,
+        P::Item: ToSql,
+        F: FnOnce(&Row) -> T {
+        self.conn.query_row(sql, params, f)
+            .optional()
+            .expect(SQL_FAIL_MESSAGE)
+    }
+
+    fn load_contract(&self, contract_name: &str) -> Option<ContractAnalysis> {
+        let result: Option<String> = self.query_row(
+            "SELECT analysis FROM type_analysis_table WHERE contract_name = ?",
+            &[contract_name],
+            |row| row.get(0));
+        match result {
+            Some(contract) => Some(ContractAnalysis::deserialize(&contract)),
```

Comment:
I'm sort-of okay with using JSON as the serialization format here, but there are two considerations:

* We need to be sure that we don't encode anything numerically large or floating-point.  JSON doesn't provide a standard way of dealing with big numbers or floats (IEEE 754 isn't even guaranteed).

* We should probably compress the JSON before storing it, since it will be verbose.

Reply:
Yeah, I'm a fan on _not_ using JSON for the eventual format. JSON is getting used here (and in the datamaps database) simply out of laziness in the expectation that we have to do some more thinking and planning around the actual final data formats, and until that time, should go with the absolute minimum viable solution, so that we can easily throw it away when the time comes.